### ISSUE TITLE
Lambda Calculus

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#8326]
Lambda Calculus<br>
<br>
https://tgdwyer.github.io/lambdacalculus/<br>
<br>
<br>
<br>
<div>
<div>
<div>
<div>
<div>
<h2>Learning Outcomes</h2>
<ul>
<li>Understand that the lambda calculus provides a complete model of computation</li>
<li>Relate the lambda calculus to functional programming</li>
<li>Apply conversion and reduction rules to simplify lambda expressions</li>
</ul>
<h2>Introduction</h2>
<p>The Lambda Calculus is a model of computation developed in the 1930s by the mathematician Alonzo Church. You are probably aware of the more famous model for computation developed around the same time by Alan Turing: the Turing Machine. However, while the Turing Machine is based on a hypothetical physical machine (involving tapes from which instructions are read and written) the Lambda Calculus was conceived as a set of rules and operations for function abstraction and application. It has been proven that, as a model of computation, the Lambda Calculus is just as powerful as Turing Machines, that is, any computation that can be modelled with a Turing Machine can also be modeled with the Lambda Calculus.</p>
<p>The Lambda Calculus is also important to study as it is the basis of functional programming. The operations we can apply to Lambda Calculus expressions to simplify (or reduce) them, or to prove equivalence, can also be applied to pure functions in a programming language that supports <a href="https://tgdwyer.github.io/higherorderfunctions">higher-order functions</a>.</p>
<h2>Lambda Expressions</h2>
<p>Lambda Calculus expressions are written with a standard system of notation. It is worth looking at this notation before studying haskell-like languages because it was the inspiration for Haskell syntax. Here is a simple Lambda Abstraction of a function:</p>
<p>The <code>&lambda;</code> (Greek letter Lambda) simply denotes the start of a function expression. Then follows a list of parameters (in this case we have only a single parameter called <code>x</code>) terminated by <code>.</code>. After the <code>.</code> is the function body, an expression returned by the function when it is applied. A variable like <code>x</code> that appears in the function body and also the parameter list is said to be <em>bound</em> to the parameter. Variables that appear in the function body but not in the parameter list are said to be <em>free</em>. The above lambda expression is equivalent to the JavaScript expression:</p>
<hr>
<h4>Exercise</h4>
<p>When we discussed combinators in JavaScript, we gave this function a name. What was it? <em><a href="https://tgdwyer.github.io/higherorderfunctions#identity-i-combinator">spoiler</a></em></p>
<hr>
<p>Some things to note about such lambda expressions:</p>
<ul>
<li>A lambda expression has no name, it is anonymous. Note that anonymous functions in languages like JavaScript and Python are also frequently called lambda expressions, or just lambdas. Now you know why.</li>
<li>The only values that Lambda Calculus variables can take on is other functions (i.e. lambda expressions). It&rsquo;s lambdas all the way down! However, to actually model and perform useful computations we say that certain expressions represent values. See the discussion of <a href="https://tgdwyer.github.io/lambdacalculus/#church-encodings">Church Encodings</a>, below, to see how this is done.</li>
<li>The names of variables bound to parameters in a lambda expression are only meaningful within the context of that expression. Thus, <code>&lambda;x.x</code> is semantically equivalent (or <em>alpha</em> equivalent) to <code>&lambda;y.y</code> or any other possible renaming of the variable.</li>
<li>Lambda functions can have multiple parameters in the parameter list, e.g.: <code>&lambda;xy. x y</code>, but they are implicitly curried (e.g. a sequence of nested univariate functions). Thus the following are all equivalent:</li>
</ul>
<div>
<div>
<pre><code>&lambda;xy.xy
= &lambda;x.&lambda;y.xy
= &lambda;x.(&lambda;y.xy)
</code></pre>
</div>
</div>
<h2>Combinators</h2>
<p>We have already discussed combinators in JavaScript, now we can give them a more formal definition:</p>
<ul>
<li>A <strong>combinator</strong> is a lambda expression (function) with no free variables.</li>
</ul>
<p>Thus, the expression <code>&lambda;x.x</code> is a combinator because the variable <code>x</code> is bound to the parameter. The expression <code>&lambda;x.xy</code> is not a combinator, because <code>y</code> is not bound to any parameter, it is <em>free</em>.</p>
<p>The <a href="https://tgdwyer.github.io/higherorderfunctions/#k-combinator"><code>K</code> combinator</a> which we wrote as <code>x=&gt;y=&gt;x</code> in JavaScript, is written <code>&lambda;xy.x</code>.</p>
<h2>Application</h2>
<p>What can we do with such a lambda expression? Well we can <em>apply</em> it to another expression (The same way we can <em>apply</em> anonymous functions to an argument in JavaScript). Here, we apply the lambda <code>(&lambda;x.x)</code> to the variable <code>y</code>:</p>
<pre><code>(&lambda;x.x)y
</code></pre>
<p>Note that while in JavaScript application of a function <code>(x=&gt;x)</code> to an argument <code>y</code> requires brackets around the argument: <code>(x=&gt;x)(y)</code>, in the Lambda Calculus application of some expression <code>f</code> to some other expression <code>x</code> is indicated simply <code>fx</code>. Brackets are required to delineate the start and end of an expression, e.g. in <code>(&lambda;x.x)y</code>, the brackets make it clear that <code>y</code> is not part of the lambda <code>&lambda;x.x</code>, but rather the lambda is being applied to <code>y</code>.</p>
<p>We can reduce this expression to a simpler form by a substitution, indicated by a bit of intermediate notation. Two types of annotations are commonly seen, you can use either (or both!):</p>
<pre><code>(&lambda;x. x) y      [x:=y]    -- an annotation on the right showing the substitution that will be applied to the expression on the left
(&lambda;x [x:=y].x)            -- an annotation inside the parameter list showing the substitution that will be performed inside the body (arguments have already been removed)
</code></pre>
<p>Now we perform the substitution in the body of the expression and throw away the head, since all the bound variables are substituted, leaving only:</p>
<p>This first reduction rule, substituting the arguments of a function application to all occurrences of that parameter inside the function body, is called <em>beta reduction</em>.</p>
<p>The next rule arises from the observation that, for some lambda term <code>M</code> that does not involve <code>x</code>:</p>
<p>is just the same as M. This last rule is called <em>eta conversion</em>.</p>
<p>Function application is left-associative except where terms are grouped together by brackets. This means that when a Lambda expression involves more than two terms, BETA reduction is applied left to right, i.e.,</p>
<p><code>(&lambda;z.z) (&lambda;a.a a) (&lambda;z.z b) = ( (&lambda;z.z) (&lambda;a.a a) ) (&lambda;z.z b)</code>.</p>
<div>
<h2>Lambda Calculus Cheatsheet</h2>
<p>Three operations can be applied to lambda expressions:</p>
<p><strong>Alpha Equivalence</strong>: variables can be arbitrarily renamed as long as the names remain consistent within the scope of the expression.</p>
<pre><code>&lambda;xy.yx = &lambda;wv.vw
</code></pre>
<p><strong>Beta Reduction</strong>: functions are applied to their arguments by substituting the text of the argument in the body of the function.</p>
<pre><code>(&lambda;x. x) y
= (&lambda;x [x:=y]. x)     - we indicate the substitution that is going to occur inside []
= x [x:=y]           - an alternative way to show the substitution
= y
</code></pre>
<p><strong>Eta Conversion</strong>: functions that simply apply another expression to their argument can be substituted with the expression in their body.</p>
<pre><code>&lambda;x.Mx
= M
</code></pre>
</div>
<p>One thing to note about the lambda calculus is that it does not have any such thing as a global namespace. All variables must be:</p>
<ul>
<li>Parameters from some enclosing lambda expression (note, below we start using labels to represent expressions - these are not variables, just placeholders for an expression that can be substituted for the label).</li>
<li>Immutable - there is no way to assign a new value to a variable from within a lambda expression.</li>
</ul>
<p>This makes the language and its evaluation very simple. All we (or any hypothetical machine for evaluating lambda expressions) can do with a lambda is apply the three basic alpha, beta and eta reduction and conversion rules. Here&rsquo;s a fully worked example of applying the different rules to reduce an expression until no more Beta reduction is possible, at which time we say it is in <em>beta normal form</em>:</p>
<div>
<div>
<pre><code>(&lambda;z.z) (&lambda;a.a a) (&lambda;z.z b)
&rArr;
((&lambda;z.z) (&lambda;a.a a)) (&lambda;z.z b)    =&gt; Function application is left-associative
&rArr;
(z [z:=&lambda;a.a a]) (&lambda;z.z b)      =&gt; BETA Reduction
&rArr;
(&lambda;a.a a) (&lambda;z.z b)
&rArr;
a a [a:=&lambda;z.z b]               =&gt; BETA Reduction
&rArr;
(&lambda;z.z b) (&lambda;z.z b)
&rArr;
z b [z:=(&lambda;z.z b)]             =&gt; BETA Reduction
&rArr;
(&lambda;z.z b) b
&rArr;
z b [z:=b]                    =&gt; BETA Reduction
&rArr;
b b         =&gt; Beta normal form, cannot be reduced again.
</code></pre>
</div>
</div>
<p>Note, sometimes I add extra spaces as above just to make things a little more readable - but it doesn&rsquo;t change the order of application, indicate a variable is not part of a lambda to its left (unless there is a bracket) or have any other special meaning.</p>
<h2>Church Encodings</h2>
<p>And yet, this simple calculus is sufficient to perform computation. Alonzo Church demonstrated that we can model any of the familiar programming language constructs with lambda expressions. For example, Booleans:</p>
<pre><code>TRUE = &lambda;xy.x    = K-combinator
FALSE = &lambda;xy.y   = K I
</code></pre>
<p>Note that we are making use of the K and I combinators here as we did for the head and rest functions for our <a href="https://tgdwyer.github.io/higherorderfunctions/#k-combinator">cons list</a>, i.e. returning either the first or second parameter to make a choice between two options. Now we can make an IF expression:</p>
<pre><code>IF = &lambda;btf.b t f
</code></pre>
<p><code>IF TRUE</code> returns the expression passed in as <code>t</code> and <code>IF FALSE</code> returns the expression passed in as <code>f</code>. Now we can make Boolean operators:</p>
<pre><code>AND = &lambda;xy. IF x  y FALSE
OR = &lambda;xy. IF x TRUE y 
NOT = &lambda;x. IF x FALSE TRUE
</code></pre>
<p>And now we can evaluate logical expressions with beta reduction:</p>
<pre><code>NOT TRUE
= (&lambda;x. IF x FALSE TRUE) TRUE       - expand NOT
= IF x FALSE TRUE [x:=TRUE]        - beta reduction
= IF TRUE FALSE TRUE
= (&lambda;btf.b t f) TRUE FALSE TRUE     - expand IF
= b t f [b:=TRUE,t:=FALSE,f:=TRUE] - beta reduction
= TRUE FALSE TRUE
= (&lambda;xy.x) FALSE TRUE               - expand TRUE
= x [x:=FALSE]                     - beta reduction
= FALSE
</code></pre>
<hr>
<h3>Exercises</h3>
<ul>
<li>Try using beta reduction to compute some more logical expressions. E.g. make an expression for XOR.</li>
<li>Our JavaScript <a href="https://tgdwyer.github.io/higherorderfunctions/#k-combinator">cons list</a> was based on the Church Encoding for linked lists. Try writing the <code>cons</code>, <code>head</code> and <code>rest</code> functions as Lambda expressions.</li>
<li>Investigate <a href="https://en.wikipedia.org/wiki/Church_encoding">Church Numerals</a> and try using lambda calculus to compute some basic math.</li>
</ul>
<hr>
<h2>Divergent Lambda Expressions</h2>
<p>Despite the above demonstration of evaluation of logical expressions, the restriction that lambda expressions are anonymous makes it a bit difficult to see how lambda calculus can be a general model for useful computation. For example, how can we have a loop? How can we have recursion if a lambda expression does not have any way to refer to itself?</p>
<p>The first hint to how loops might be possible with lambda calculus is the observation that some expressions do not simplify when beta reduced. For example:</p>
<pre><code>( &lambda;x . x  x) ( &lambda;y. y y)   - (1)
x x [x:= y. y y]
( &lambda;y . y  y) ( &lambda;y. y y)   - which is alpha equivalent to what we started with, so goto (1)
</code></pre>
<p>Thus, the reduction would go on forever. Such an expression is said to be divergent. However, if a lambda function is not able to refer to itself it is still not obvious how recursion is possible.</p>
<p>The answer is due to the American mathematician Haskell Curry and is called the fixed-point or Y combinator:</p>
<pre><code> Y = &lambda;f. ( &lambda;x . f (x x) ) ( &lambda;x. f (x x) )
</code></pre>
<p>When we apply <code>Y</code> to another function <code>g</code> we see an interesting divergence:</p>
<pre>
Y g = (&lambda;f. ( &lambda;x . f (x x) ) ( &lambda;x. f (x x) ) ) g
    = ( &lambda;x . f (x x) ) ( &lambda;x. f (x x) ) [f:=g]  <i>- beta reduction</i>
    = <b>( &lambda;x . g (x x) ) ( &lambda;x. g (x x) )</b>         <i>- a partial expansion of Y g, remember this...</i>
    = g (x x) [ x:= &lambda;x. g (x x)]               <i>- beta reduction</i>
    = g ( <b>(&lambda;x. g (x x) (&lambda;x. g (x x) )</b> )        <i>- bold part matches Y g above, so now...</i>
    = g (Y g)
<i>... more beta reduction as above
... followed by substitution with Y g when we see the pattern above...</i>
    = g (g (Y g))
    = g (g (g (Y g)))
<i>... etc</i>
</pre>
<p>If we directly translate the above version of the Y-combinator into JavaScript we get the following:</p>
<div>
<div>
<pre><code><span>const</span> <span>Y</span> <span>=</span> <span>f</span><span>=&gt;</span> <span>(</span><span>x</span> <span>=&gt;</span> <span>f</span><span>(</span><span>x</span><span>(</span><span>x</span><span>)))(</span><span>x</span><span>=&gt;</span> <span>f</span><span>(</span><span>x</span><span>(</span><span>x</span><span>)))</span> <span>// warning infinite recursion ahead!</span>
</code></pre>
</div>
</div>
<p>So now <code>Y</code> is just a function which can be applied to another function, but what sort of function do we pass into <code>Y</code>? If we are to respect the rules of the Lambda calculus we cannot have a function that calls itself directly. That is, because Lambda expressions have no name, they can&rsquo;t refer to themselves by name.</p>
<p>Therefore, we need to wrap the recursive function in a lambda expression into which a reference to the recursive function itself can be passed as parameter. We can then in the body of the function refer to the parameter function by name. It&rsquo;s a bit weird, let me just give you a JavaScript function which fits the bill:</p>
<div>
<div>
<pre><code><span>// A function that recursively calculates 'n!'</span>
<span>// - but it needs to be reminded of its own name in the f parameter in order to call itself.</span>
<span>const</span> <span>FAC</span> <span>=</span> <span>f</span> <span>=&gt;</span> <span>n</span> <span>=&gt;</span> <span>n</span><span>&gt;</span><span>1</span> <span>?</span> <span>n</span> <span>*</span> <span>f</span><span>(</span><span>n</span><span>-</span><span>1</span><span>)</span> <span>:</span> <span>1</span>
</code></pre>
</div>
</div>
<p>Now we can make this function compute factorials like so:</p>
<div>
<div>
<pre><code><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>())))))(</span><span>6</span><span>)</span>
</code></pre>
</div>
</div>
<blockquote>
<p>720</p>
</blockquote>
<p>Because we gave FAC a stopping condition, we can call too many times and it will still terminate:</p>
<div>
<div>
<pre><code><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>(</span><span>FAC</span><span>()))))))))(</span><span>6</span><span>)</span>
</code></pre>
</div>
</div>
<blockquote>
<p>720</p>
</blockquote>
<p>From the expansion of <code>Y g = g (g (g (...)))</code> it would seem that <code>Y(FAC)</code> would give us the recurrence we need. But will the JavaScript translation of the Y-combinator be able to generate this sequence of calls?</p>
<blockquote>
<p>stack overflow</p>
</blockquote>
<p>Well we got a recurrence, but unfortunately the JavaScript engine&rsquo;s strict (or eager) evaluation means that we must completely evaluate Y(FAC) before we can ever apply the returned function to (6).<br>
Therefore, we get an infinite loop - and actually it doesn&rsquo;t matter what function we pass in to Y, it will never actually be called and any stopping condition will never be checked. How do we restore the laziness necessary to make progress in this recursion?</p>
<p>(<strong>Hint:</strong> it involves wrapping some part of <code>Y</code> in another lambda)</p>
<p>Did you get it? If so, good for you! If not, never mind, it is tricky and in fact was the subject of research papers at one time, so I&rsquo;ll give you a bigger hint.</p>
<p><strong>Bigger hint:</strong> there&rsquo;s another famous combinator called <code>Z</code> which is basically <code>Y</code> adapted to work with strict evaluation:</p>
<pre><code>Z=&lambda;f.(&lambda;x.f(&lambda;v.xxv))(&lambda;x.f(&lambda;v.xxv))
</code></pre>
<hr>
<h3>Exercises</h3>
<ul>
<li>Note the similarities between <code>Y</code> and <code>Z</code> and perform a similar set of Beta reductions on <code>Z FAC</code> to see how it forces FAC to be evaluated.</li>
<li>Write a version of the Z-Combinator in JavaScript such that <code>Z(FAC)(6)</code> successfully evaluates to <code>720</code>.</li>
</ul>
<hr>
<h2>Conclusion</h2>
<p>If you want to dig deeper there is much <a href="https://www.seas.harvard.edu/courses/cs152/2015sp/lectures/lec07-encodings.pdf">more written about Lambda Calculus encodings</a> of logical expressions, natural numbers, as well as the <code>Y</code> and <code>Z</code> combinators, and also <a href="https://blog.benestudio.co/fixed-point-combinators-in-javascript-c214c15ff2f6">more about their implementation in JavaScript</a>.</p>
<p>However, the above description should be enough to give you a working knowledge of how to apply the three operations to manipulate Lambda Calculus expressions, as well as an appreciation for how they can be used to reason about combinators in real-world functional style curried code. The other important take away is that the Lambda Calculus is a turing-complete model of computation, with Church encodings demonstrating how beta-reduction can evaluate church-encoded logical and numerical expressions and the trick of the Y-combinator giving us a way to perform loops.</p>
</div>
</div>
</div>
</div>
</div>
<br>
<br>
<br>
<br>
<br>
<br>
via Tim&rsquo;s code stuff<br>
<br>
February 14, 2022 at 04:54PM
